source/common/http/alternate_protocols_cache_impl.cc

```diff
@@ -0,0 +1,51 @@
+#include "common/http/alternate_protocols_cache_impl.h"
+
+namespace Envoy {
+namespace Http {
+
+AlternateProtocolsCacheImpl::AlternateProtocolsCacheImpl(ThreadLocal::Instance& tls,
+                                                         TimeSource& time_source)
+    : time_source_(time_source), slot_(tls) {
+  slot_.set([](Event::Dispatcher& /*dispatcher*/) { return std::make_shared<State>(); });
+}
+
+AlternateProtocolsCacheImpl::~AlternateProtocolsCacheImpl() {}
+
+void AlternateProtocolsCacheImpl::setAlternatives(const Origin& origin,
+                                                  const std::vector<AlternateProtocol>& protocols,
+                                                  const MonotonicTime& expiration) {
+  // TODO(RyanTheOptimist): Propogate state changes across threads.
+  State& state = *(slot_.get());
+  Entry& entry = state.protocols_[origin];
+  if (entry.protocols_ != protocols) {
+    entry.protocols_ = protocols;
+  }
+  if (entry.expiration_ != expiration) {
+    entry.expiration_ = expiration;
+  }
+}
+
+OptRef<const std::vector<AlternateProtocolsCache::AlternateProtocol>>
+AlternateProtocolsCacheImpl::findAlternatives(const Origin& origin) {
+  State& state = *(slot_.get());
+  auto entry_it = state.protocols_.find(origin);
+  if (entry_it == state.protocols_.end()) {
+    return makeOptRefFromPtr<const std::vector<AlternateProtocol>>(nullptr);
+  }
+
+  const Entry& entry = entry_it->second;
+  if (time_source_.monotonicTime() > entry.expiration_) {
+    // Expire the entry.
```

Comment:
I'm noticing that we don't ever expire entries that aren't being queried for. Add a TODO to do that?

Reply:
Done. FWIW, Chrome's HttpServerProperties doesn't actually expires old entries on a timer but we can do better here! :)

Reply:
Yeah Envoy probably has a much longer process lifetime than Chrome being server-side and all :)

And BTW a timer isn't the only approach, you could check the list any time a query comes in. And if you want to go full on interview question you could use a data structure that sorts by expiration to make that cheaper :P 

But yeah a TODO sounds good :)